### PR TITLE
Endpoint for local node info

### PR DESCRIPTION
**Pull request by kaloyan-raev**
Fixes https://storjlabs.atlassian.net/browse/V3-1240

Adds a new `Info` method to the Kademlia endpoint that returns the following local node info:
- ID
- Type
- Metadata (email and wallet)
- Restrictions (free storage and bandwidth)

The new endpoint is exposed as `inspector kad node-info` command too.

### CODE REVIEW

File: pkg/kademlia/dialer.go [hunk 1]
```diff
@@ -109,6 +109,23 @@ func (dialer *Dialer) FetchPeerIdentity(ctx context.Context, target pb.Node) (pI
 	return identity.PeerIdentityFromPeer(p)
 }
 
+// Info connects to a node address and returns its node info.
+func (dialer *Dialer) Info(ctx context.Context, address *pb.NodeAddress) (*pb.InfoResponse, error) {
```
Comment (egonelbre) [inline pkg/kademlia/dialer.go]: Just realizing that having `Dialer` with methods `FetchPeerIdentity` and `Info` seems weird, seems like it's being done this way because it's convenient rather than clear separation of concerns. `Dialer` was meant mainly for handling kademlia communication protocol, nothing else. We can think how to handle that later.

Also instead of `Info`, maybe use method name `FetchInfo` to make clearer that it's making a request.

File: pkg/kademlia/dialer.go [hunk 2]
```diff
@@ -109,6 +109,23 @@ func (dialer *Dialer) FetchPeerIdentity(ctx context.Context, target pb.Node) (pI
...
+		return nil, context.Canceled
+	}
+	defer dialer.limit.Unlock()
+
+	conn, err := dialer.dialAddress(ctx, address)
+	if err != nil {
+		return nil, err
+	}
+
+	resp, err := conn.client.Info(ctx, &pb.InfoRequest{})
```
Comment (egonelbre) [inline pkg/kademlia/dialer.go]: Maybe `client.RequestInfo`? The GRPC methods seem to be verbs rather than nouns.

File: pkg/pb/overlay.proto [hunk 3]
```diff
@@ -85,6 +86,16 @@ message QueryResponse {
 message PingRequest {};
 message PingResponse {};
 
+// TODO: add fields that validate who is requesting the info
+message InfoRequest {}
+
+message InfoResponse {
+    bytes id = 1 [(gogoproto.customtype) = "NodeID", (gogoproto.nullable) = false];
```
Comment (egonelbre) [inline pkg/pb/overlay.proto]: This shouldn't be here. We cannot trust the other side giving valid response. Use `PeerIdentityFromPeer` to get the identity.

File: pkg/pb/overlay.proto [hunk 4]
```diff
@@ -85,6 +86,16 @@ message QueryResponse {
...
+    node.NodeType type = 2;
+    node.NodeMetadata metadata = 3;
+    node.NodeRestrictions restrictions = 4;
```
Comment (egonelbre) [inline pkg/pb/overlay.proto]: Is there a better name for this?

File: pkg/pb/overlay.proto [hunk 5]
```diff
@@ -85,6 +86,16 @@ message QueryResponse {
 message PingRequest {};
 message PingResponse {};
 
+// TODO: add fields that validate who is requesting the info
+message InfoRequest {}
+
+message InfoResponse {
+    bytes id = 1 [(gogoproto.customtype) = "NodeID", (gogoproto.nullable) = false];
+    node.NodeType type = 2;
+    node.NodeMetadata metadata = 3;
```
Comment (egonelbre) [inline pkg/pb/overlay.proto]: Is there a better name for this? `NodeOperator` maybe?

File: pkg/pb/overlay.proto (hunk 4 above)
Comment (kaloyan-raev) [reply pkg/pb/overlay.proto]: Maybe there is a better name, but `NodeRestrictions` is an already defined type from the past. It's just referenced here.

File: pkg/pb/overlay.proto (hunk 5 above)
Comment (kaloyan-raev) [reply pkg/pb/overlay.proto]: Same here. The `NodeMetadata` is an already existing type. It is not defined in this PR.

File: pkg/pb/overlay.proto (hunk 4 above)
Comment (egonelbre) [reply pkg/pb/overlay.proto]: I'm aware that these are already defined, it's more of we can improve the naming more easily right now, instead of later when it's integrated into elsewhere.

File: pkg/pb/overlay.proto (hunk 5 above)
Comment (egonelbre) [reply pkg/pb/overlay.proto]: I'm aware that these are already defined, it's more of we can improve the naming more easily right now, instead of later when it's integrated into elsewhere.

File: pkg/pb/overlay.proto (hunk 4 above)
Comment (egonelbre) [reply pkg/pb/overlay.proto]: Also completely fine to handle it later.

File: pkg/pb/overlay.proto (hunk 5 above)
Comment (egonelbre) [reply pkg/pb/overlay.proto]: Also completely fine to handle it later.

File: pkg/pb/overlay.proto (hunk 3 above)
Comment (kaloyan-raev) [reply pkg/pb/overlay.proto]: I don't understand. The ID is just one of the "info" properties returned by the node. So if the sender has just the address of the target node, it will receive the target's ID, type, metadata and restrictions. I don't understand how this relates with the peer identity.

Comment (egonelbre) [reply pkg/pb/overlay.proto]: Because the responder can write anything it wants to that field.

Comment (egonelbre) [reply pkg/pb/overlay.proto]: The only thing we can trust is the NodeID and certificate that it's using for communicating with us. Because, we can verify that it actually has the corresponding private key and certificate.

File: pkg/kademlia/dialer.go (hunk 1 above)
Comment (kaloyan-raev) [reply pkg/kademlia/dialer.go]: Changed in latest commit.

File: pkg/kademlia/dialer.go (hunk 2 above)
Comment (kaloyan-raev) [reply pkg/kademlia/dialer.go]: Changed in latest commit.

File: pkg/pb/overlay.proto (hunk 3 above)
Comment (kaloyan-raev) [reply pkg/pb/overlay.proto]: OK. I see the issue. I will remove the ID from the response.

Comment (kaloyan-raev) [reply pkg/pb/overlay.proto]: Done.

File: pkg/pb/overlay.proto (hunk 5 above)
Comment (kaloyan-raev) [reply pkg/pb/overlay.proto]: OK. I will define a new `NodeOperator` here and remove `NodeMetadata` in #1341.

File: pkg/pb/overlay.proto (hunk 4 above)
Comment (kaloyan-raev) [reply pkg/pb/overlay.proto]: How about `NodeCapacity` instead of `NodeRestrictions`?

Comment (egonelbre) [reply pkg/pb/overlay.proto]: Yes, that's a really nice name.

File: pkg/kademlia/dialer.go [hunk 6]
```diff
@@ -109,6 +109,23 @@ func (dialer *Dialer) FetchPeerIdentity(ctx context.Context, target pb.Node) (pI
...
+	defer dialer.limit.Unlock()
+
+	conn, err := dialer.dial(ctx, target)
+	if err != nil {
+		return nil, err
+	}
+
+	resp, err := conn.client.RequestInfo(ctx, &pb.InfoRequest{})
+
+	return resp, errs.Combine(err, conn.disconnect())
```
Comment (egonelbre) [inline pkg/kademlia/dialer.go]: Since we established connection we can still return the node id. So the method would be:

```
func (dialer *Dialer) FetchInfo(ctx context.Context, target pb.Node) (storj.NodeID, *pb.InfoResponse, error)

    p  := &peer.Peer{}
    pCall := grpc.Peer(p)
    resp, err = conn.client.RequestInfo(ctx, &pb.InfoRequest{}, pCall)
    id ... := identity.PeerIdentityFromPeer(p)
   return id, resp, ...
}
```

Comment (kaloyan-raev) [reply pkg/kademlia/dialer.go]: OK. This sounds good. I will bring the Node ID back to the response.

File: pkg/pb/overlay.proto (hunk 5 above)
Comment (kaloyan-raev) [reply pkg/pb/overlay.proto]: Done.

File: pkg/pb/overlay.proto (hunk 4 above)
Comment (kaloyan-raev) [reply pkg/pb/overlay.proto]: Done.

File: pkg/pb/inspector.proto [hunk 7]
```diff
@@ -110,6 +112,17 @@ message LookupNodeResponse {
   node.NodeMetadata meta = 2;
 }
 
+message NodeInfoRequest {
+  bytes id = 1 [(gogoproto.customtype) = "NodeID", (gogoproto.nullable) = false];
```
Comment (egonelbre) [inline pkg/pb/inspector.proto]: I'm confused. Why is this here?

Comment (egonelbre) [reply pkg/pb/inspector.proto]: ah, wait this is the kademlia thing.

File: pkg/kademlia/dialer.go (hunk 6 above)
Comment (kaloyan-raev) [reply pkg/kademlia/dialer.go]: Done.

File: pkg/pb/inspector.proto (hunk 7 above)
Comment (kaloyan-raev) [reply pkg/pb/inspector.proto]: Initially it was only the address in the `NodeInfoRequest`. Then, when you asked to remove the ID from the response, I added the ID in the `NodeInfoRequest`, just for the sake for looking like other similar requests - `PingRequest` and `FetchPeerIdentity`.

I removed it again in the latest commit, because it does not makes sense to me. Having only the address is enough to complete the `NodeInfoRequest`. This may be helpful for other use cases where the caller knows only the address of the bootstrap node or the satellite and wants to find its ID using the `NodeInfoRequest`.

File: pkg/kademlia/dialer.go [hunk 8]
```diff
@@ -109,6 +109,26 @@ func (dialer *Dialer) FetchPeerIdentity(ctx context.Context, target pb.Node) (pI
...
+
+	conn, err := dialer.dialAddress(ctx, address)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	p := &peer.Peer{}
+	pCall := grpc.Peer(p)
+	resp, err := conn.client.RequestInfo(ctx, &pb.InfoRequest{}, pCall)
+	id, idErr := identity.PeerIdentityFromPeer(p)
```
Comment (mniewrzal) [inline pkg/kademlia/dialer.go]: Should we do this step if `RequestInfo` returns error?

Comment (kaloyan-raev) [reply pkg/kademlia/dialer.go]: You are right. Fixed it in the latest commit.
